python/lsst/pipe/base/connections.py

```diff
@@ -607,3 +629,111 @@ def iterConnections(connections: PipelineTaskConnections,
+    translating them to and from the connection-oriented mappings used inside
+    `PipelineTaskConnections`.
+    """
+
+    def adjust_in_place(
+        self,
+        connections: PipelineTaskConnections,
+        label: str,
+        data_id: DataCoordinate,
+    ) -> None:
+        """Call `~PipelineTaskConnections.adjustQuantum` and update ``self``
+        with its results.
+
+        Parameters
+        ----------
+        connections : `PipelineTaskConnections`
+            Instance on which to call `~PipelineTaskConnections.adjustQuantum`.
+        label : `str`
+            Label for this task in the pipeline (should be used in all
+            diagnostic messages).
+        data_id : `lsst.daf.butler.DataCoordinate`
+            Data ID for this quantum in the pipeline (should be used in all
+            diagnostic messages).
+        """
+        # Translate self's DatasetType-keyed, Quantum-oriented mappings into
+        # connection-keyed, PipelineTask-oriented mappings.
+        inputs_by_connection: typing.Dict[str, typing.Tuple[BaseInput, typing.Tuple[DatasetRef, ...]]] = {}
+        outputs_by_connection: typing.Dict[str, typing.Tuple[Output, typing.Tuple[DatasetRef, ...]]] = {}
+        for name in itertools.chain(connections.inputs, connections.prerequisiteInputs):
+            connection = getattr(connections, name)
+            dataset_type_name = connection.name
+            inputs_by_connection[name] = (
+                connection,
+                tuple(self.inputs.get(dataset_type_name, ()))
+            )
+        for name in itertools.chain(connections.outputs):
+            connection = getattr(connections, name)
+            outputs_by_connection[name] = (
+                connection,
+                tuple(self.outputs.get(dataset_type_name, ()))
+            )
+        # Actually call adjustQuantum.
+        adjusted_inputs_by_connection, adjusted_outputs_by_connection = connections.adjustQuantum(
+            inputs_by_connection,
+            outputs_by_connection,
+            label,
+            data_id,
+        )
+        # Translate adjustments to DatasetType-keyed, Quantum-oriented form,
+        # installing new mappings in self if necessary.
+        adjusted_inputs: typing.Optional[NamedKeyDict[DatasetType, typing.List[DatasetRef]]] = None
+        adjusted_outputs: typing.Optional[NamedKeyDict[DatasetType, typing.List[DatasetRef]]] = None
+        for name, (connection, updated_refs) in adjusted_inputs_by_connection.items():
+            dataset_type_name = connection.name
+            if not set(updated_refs).issubset(self.inputs[dataset_type_name]):
+                raise RuntimeError(
+                    f"adjustQuantum implementation for task with label {label} returned {name} "
+                    f"({dataset_type_name}) input datasets that are not a subset of those "
+                    f"it was given for data ID {data_id}."
+                )
+            if adjusted_inputs is None:
```

Comment:
Why not just do this above? It seems like it will happen everytime the function is called anyway, unless the exception is raised above, but it does not seem to hurt in that case

Reply:
(Premature?) optimization: just didn't want to make a `NamedKeyDict` instance at all if there are no adjustments, since that's the common case.

Reply:
Oh I see you are relying on the Mapping to be empty, and no looping to actually happen and set up and change a sentinel and then change it in the loop if it something happens.

Since you are Doing and explicit if block below anyway, might I suggest changing the structure to be:
```
if adjusted_inputs_by_connection:
    adjusted_inputs = NamedKeyDict()
    for name, (connection, updated_refs in adjusted_inputs_by_connection.items():
        ...
        adjusted_inputs[dataset_type_name] = list(updated_refs
    self.inputs = adjusted_inputs.freeze()
    self.inputs_adjusted = True
else:
    # do you need an else? shouldn't it default to false? I guess you may keep a helper around for a while, but then I
    # question if you can do re-use of instances anyway how do you handle if something was true, becomes false, etc
    self.inputs_adjusted = Flase
    
```
same for the adjusted_outputs_by_connection block